Mathlib/GroupTheory/DivisibleHull.lean

```diff
@@ -0,0 +1,411 @@
+    a • mk m s = mk (a.num • m) (⟨a.den, a.den_pos⟩ * s) := by
+  obtain h | h := le_total 0 a
+  · rw [qsmul_of_nonneg h, nnqsmul_mk, ← natCast_zsmul]
+    congr
+    simpa using h
+  · rw [qsmul_of_nonpos h]
+    have : a.num.natAbs • m = -a.num • m := by
+      rw [← natCast_zsmul]
+      congr
+      simpa using h
+    simp [nnqsmul_mk, this, ← neg_mk]
+
+noncomputable
+instance : Module ℚ (DivisibleHull M) where
+  one_smul x := by
+    induction x with | mk m s
+    simp [qsmul_of_nonneg zero_le_one, nnqsmul_mk]
+  zero_smul := zero_qsmul
+  smul_zero a := by simp [qsmul_def]
+  smul_add a x y := by simp [qsmul_def, smul_add]
+  add_smul a b x := by
+    induction x with | mk m s
+    simp_rw [qsmul_mk, mk_add_mk, mk_eq_mk]
+    use 1
+    suffices ((a + b).num * a.den * b.den * (s * s)) • m =
+        ((a.num * b.den + b.num * a.den) * (a + b).den * (s * s)) • m  by
+      convert this using 1
+      all_goals
+      simp [← natCast_zsmul, smul_smul, ← add_smul]
+      ring_nf
+    rw [Rat.add_num_den']
+  mul_smul a b x := by
+    induction x with | mk m s
+    simp_rw [qsmul_mk, mk_eq_mk]
+    use 1
+    suffices ((a * b).num * a.den * b.den * s) • m = (a.num * b.num * (a * b).den * s) • m by
+      convert this using 1
+      all_goals
+      simp [← natCast_zsmul, smul_smul]
+      ring_nf
+    rw [Rat.mul_num_den']
+
+theorem zsmul_mk (a : ℤ) (m : M) (s : ℕ+) : a • mk m s = mk (a • m) s := by
+  simp [← Int.cast_smul_eq_zsmul ℚ a, qsmul_mk]
+
+end Group
+
+section LinearOrder
+variable {M : Type*} [AddCommMonoid M] [LinearOrder M] [IsOrderedCancelAddMonoid M]
+
+private theorem lift_aux (m n m' n' : M) (s t s' t' : ℕ+)
+    (h : mk m s = mk m' s') (h' : mk n t = mk n' t') :
+    (t.val • m ≤ s.val • n) = (t'.val • m' ≤ s'.val • n') := by
+  rw [mk_eq_mk_iff_smul_eq_smul] at h h'
+  rw [propext_iff, ← (nsmul_right_strictMono (mul_ne_zero s'.ne_zero t'.ne_zero)).le_iff_le]
+  convert (nsmul_right_strictMono (M := M) (mul_ne_zero s.ne_zero t.ne_zero)).le_iff_le using 2
+  · simp_rw [smul_smul, mul_rotate s'.val, ← smul_smul, h, smul_smul]
+    ring_nf
+  · simp_rw [smul_smul, ← mul_rotate s'.val, ← smul_smul, ← h', smul_smul]
+    ring_nf
+instance : LE (DivisibleHull M) where
+  le x y := liftOn₂ x y (fun m s n t ↦ t.val • m ≤ s.val • n) lift_aux
+
+@[simp]
+theorem mk_le_mk {m m' : M} {s s' : ℕ+} :
+    mk m s ≤ mk m' s' ↔ s'.val • m ≤ s.val • m' := by rfl
+
+instance : LinearOrder (DivisibleHull M) where
+  le_refl a := by
+    induction a with | mk m s
+    simp
+  le_trans a b c hab hbc := by
+    induction a with | mk ma sa
+    induction b with | mk mb sb
+    induction c with | mk mc sc
+    rw [mk_le_mk] at ⊢ hab hbc
+    rw [← (nsmul_right_strictMono (show sb.val ≠ 0 by simp)).le_iff_le, smul_comm _ _ ma,
+      smul_comm _ _ mc]
+    rw [← (nsmul_right_strictMono (show sc.val ≠ 0 by simp)).le_iff_le, smul_comm _ _ mb] at hab
+    rw [← (nsmul_right_strictMono (show sa.val ≠ 0 by simp)).le_iff_le] at hbc
```

Comment:
Shouldn't you be using `nsmul_le_nsmul_iff_left` here?

Reply:
Let me quickly add it